runtime/src/main/java/org/corfudb/util/Utils.java

```diff
@@ -119,6 +117,80 @@ public static String toReadableId(UUID id) {
         return Long.toHexString((id.getLeastSignificantBits()) & 0xFFFF);
     }
 
+
+    /**
+     * Verify that the stripes are not empty and that each segment has one stripe.
+     * @param segments segments to validate
+     */
+    private static void validateSegments(List<Layout.LayoutSegment> segments) {
+        checkNotNull(segments);
+        checkArgument(!segments.isEmpty());
+        long previousSegmentEndAddress = 0;
+        for (Layout.LayoutSegment segment : segments) {
+            checkState(segment.getStart() == previousSegmentEndAddress);
+            previousSegmentEndAddress = segment.getEnd();
+            // only supported for chain replication
+            checkArgument(segment.getReplicationMode() == Layout.ReplicationMode.CHAIN_REPLICATION);
+            // Since stripping is not supported, we can assume only one stripe exists
+            checkArgument(segment.getStripes().size() == 1);
+            // A stripe cannot be empty
+            checkArgument(!segment.getStripes().get(0).getLogServers().isEmpty());
+        }
+        // The last segment (i.e open segment) end address is -1 (denoting infinity)
+        checkState(previousSegmentEndAddress == -1);
+    }
+
+    /**
+     * This method selects a head node: the first node (in the open segment) that also exists in all previous
+     * segments.
+     * @param layout layout to use to find the head node
+     * @return head node
+     * @throws IllegalStateException if no such node exists
+     */
+    private static String selectHeadNode(Layout layout) {
+        checkNotNull(layout);
+        validateSegments(layout.getSegments());
+        List<Layout.LayoutSegment> segments = layout.getSegments();
+
+        // select open segment (i.e. last segment)
+        Layout.LayoutSegment openSegment = segments.get(segments.size() - 1);
+        // only support chain replication
+        Layout.LayoutStripe openSegmentStripe = openSegment.getStripes().get(0);
+
+        if (segments.size() == 1) {
+            // Since there is only one segment we can return the first node in the stripe (i.e. head node)
+            return segments.get(0).getStripes().get(0).getLogServers().get(0);
+        }
+
+        // walk the chain of the last segment and find the first node that exists in all previous segments
+        List<Layout.LayoutSegment> reversedSegmentPrefix = segments.subList(0, segments.size() - 1);
+        checkState(!reversedSegmentPrefix.isEmpty());
```

Comment:
It's hard to understand what this check does.
What's the motivation for:
 - checking that `segments` list contains more than one element if we already checked that condition in the previous statement (a couple of lines above)?
 - reverse `reversedSegmentPrefix` collection two times, which gives us the same collection in the end

Cognitive complexity of this method seems unnecessary complicated

Reply:
Sorry confused with the name of `reversedSegmentPrefix `

Reply:
I'm trying to find the first node from the last segment that also exists in all segments. Which is different from finding a node that exists in all segments. just consider the case where you have all nodes in all segments. Then obviously not all nodes can be the head segment. 

Actually, although this algorithm is better than what we have on master, it is still in **incorrect.** It's not possible to compute the sequencer state from only **one** node. I think.

Reply:
Regarding current code, I played a bit with the code, it seems to me it can be simplified to
```
LayoutStripe openSegmentStripe = layout.getLatestSegment().getFirstStripe();

        for (String node : openSegmentStripe.getLogServers()) {
            boolean nodeIsPresent = segments.stream()
                    .allMatch(currentSegment -> currentSegment.getAllLogServers().contains(node));

            if (nodeIsPresent) {
                return node;
            }
        }
```
If it would work for you, then great, if it's not correct then ok, no worries

Reply:
Changed algorithm. Please see, the logic is cleaner #2715